### DISCUSSION TITLE
How to operate the new Popup page (ChromiumHostControl)?

### DISCUSSION THREAD

**Discussion by OdinTemple** [cefsharp/CefSharp#3898] [General]
Notice that in the Cefsharp.winforms.example, there're two AddTab methods, one for popup and the other for traditional new tab(e.g. click a link with pressing Shift key ). The previous one, handled a ` CefSharp.WinForms.Host.ChromiumHostControl` object while the other one creates `new ChromiumWebBrowser()`. But only the `ChromiumWebBrowser` object can do a lot work, so How can I get the new popup page's internal browser reference for,  for instance, run scripts, get source codes, etc.? I can't use `ChromiumHostControl as ChromiumWebBrowser` . In the example, the popup window can't read source code or zoom due to this.

**Comment by amaitland** [comment]
> How can I get the new popup page's internal browser reference for, for instance, run scripts, get source codes, etc.?

Currently you'd have to extend [LifeSpanHandler](http://cefsharp.github.io/api/95.7.x/html/T_CefSharp_WinForms_Handler_LifeSpanHandler.htm) and override [OnAfterCreated](http://cefsharp.github.io/api/95.7.x/html/M_CefSharp_WinForms_Handler_LifeSpanHandler_OnAfterCreated.htm) to get a reference to the `IBrowser` instance. You can get access to frames, browser host etc from there.

> I can't use `ChromiumHostControl as ChromiumWebBrowser`

Correct. The popups use `ChromiumHostControl` directly as it's currently a simple parent container.

I've added some changes that will be in version `96`. 

- https://github.com/cefsharp/CefSharp/commit/129879cd20c5c5b5d04d96a491fa4da4b6bd555f
- https://github.com/cefsharp/CefSharp/commit/6e9e4a90a22422d52ebde8698385e0f70a9f564c?diff=unified

Pull request with further enhancements would be welcome.

**Reply by OdinTemple** [reply]
@amaitland  I use my method: to extend  `LifeSpanHandler ` as `*myHandler*` and overrided `OnBeforePopup`

                protected override bool OnBeforePopup(IWebBrowser chromiumWebBrowser, IBrowser browser, IFrame frame, string targetUrl, string targetFrameName, WindowOpenDisposition targetDisposition, bool userGesture, IPopupFeatures popupFeatures, IWindowInfo windowInfo, IBrowserSettings browserSettings, ref bool noJavascriptAccess, out IWebBrowser newBrowser)
                {
               
                newBrowser = (ChromiumWebBrowser)chromium2;
                
                if (onPopupCreated == null)
                {
                    return false;
                }

                var webBrowser = (ChromiumWebBrowser)chromium2;

                //We need to execute sync here so IWindowInfo.SetAsChild is called before we return false;
                webBrowser.InvokeSyncOnUiThreadIfRequired(new Action(() =>
                {
                    var control = new ChromiumHostControl
                    {
                        Dock = DockStyle.Fill
                    };
                    control.CreateControl();

                    onPopupCreated?.Invoke(control, targetUrl);

                    var rect = control.ClientRectangle;

                    var windowBounds = new CefSharp.Structs.Rect(rect.X, rect.Y, rect.Width, rect.Height);

                    windowInfo.SetAsChild(control.Handle, windowBounds);
                }));

                //here, I can get the reference, but it is still created by myself, not naturally
                pop = webBrowser;
                return false;
            }



 This is how I initialized chromium2:

            chromium2 = new ChromiumWebBrowser("");
            chromium2.CreateControl();
            panel1.Controls.Add(chromium2);
            chromium2.Dock = DockStyle.Fill;
            chromium2.Load("about:blank");
            chromium2.Focus();

           
            var handler = new myHandler();
            handler.OnPopupCreated((ctrl, targetUrl) =>
            {
                //Don't try using ctrl.FindForm() here as
                //the control hasn't been attached to a parent yet.
                ctrl.Dock = DockStyle.Fill;
                ctrl.SuspendLayout();
                panel1.Controls.Clear();
                panel1.Controls.Add(ctrl);
                ctrl.ResumeLayout(true);
                tabControl1.SelectedIndex = 1;
            })
            .OnPopupDestroyed((ctrl, popupBrowser) =>
            {
                if (!ctrl.IsDisposed && ctrl.IsHandleCreated)
                {
                    ctrl.Dispose();
                }
            });
            Global.browser.LifeSpanHandler = handler;
            Global.browser.LifeSpanHandler = new  LifeSpanHandler();
            tabPage1.Controls.Add(Global.browser);
            Global.browser.Dock = DockStyle.Fill;
            Global.browser.Load(url);`


I trid many many combinations and it seems to be the only way that I can open the popup in the second tabpage and can refer to it via the ChromiumWebBrowser variable `pop`, but, all these seems strange..So I am not sure If I have done it correctly or not?

And, one more question, I read the sourceCode, I still can't find the connection between chromium2 and ChromiumHostControl?

**Reply by amaitland** [reply]
You can test out the latest 96-pre release see https://github.com/cefsharp/CefSharp/issues/3902#issuecomment-978820002

> ChromiumWebBrowser variable pop, but, all these seems strange..So I am not sure If I have done it correctly or not?

This option is experimental and should only be used in WPF. 

You don't need a ChromiumWebBrowser instance for your popup, you just need to get the IBrowser instance which is accessible in 96-pre. See https://github.com/cefsharp/CefSharp/commit/6e9e4a90a22422d52ebde8698385e0f70a9f564c?diff=unified#diff-5ce832ad0887b4ba2a2fb5a8d34fe775f48b4ccd137d6031f3dcde9cb5e7339b

> And, one more question, I read the sourceCode, I still can't find the connection between chromium2 and ChromiumHostControl?

What do you mean?

**Reply by OdinTemple** [reply]
> you just need to get the IBrowser instance which is accessible in 96-pre

but IBrowser is not usable for **any** operation, isn't it?Which IWebBrowser does. Actually I don't even know what I can do with IBrowser. IBrowser only has some properties like isPopup/canGoBack..it doesn't have methods. Did you change it?
I am sorry that I don't know how to build the pre-version and I don't know how to develop WPF for now, I only know install it via nuget so I guess I couldn't test it for now..


>And, one more question, I read the sourceCode, I still can't find the connection between chromium2 and ChromiumHostControl?

I meant, for instance, there is a ChromiumWebBrowser object and a ChromiumHostControl object connected, and I don't know what does this ChromiumHostControl is connected to the brwoser, like, if there's a private field of the ChromiumHostControl that refers to a chromium webBrowser or IWebBrowser. As you see this method:

                webBrowser.InvokeSyncOnUiThreadIfRequired(new Action(() =>
                {
                var control = new ChromiumHostControl
                {
                    Dock = DockStyle.Fill
                };
                control.CreateControl();

                onPopupCreated?.Invoke(control, targetUrl);

                var rect = control.ClientRectangle;

                var windowBounds = new CefSharp.Structs.Rect(rect.X, rect.Y, rect.Width, rect.Height);

                windowInfo.SetAsChild(control.Handle, windowBounds);
            }));

It looks like that I created a solitary ChromiumHostControl object and it just suddenly related to the webBrowser.

**Reply by amaitland** [reply]
> but IBrowser is not usable for **any** operation, isn't it?

Pretty much everything is just an extension method based on `IBrowser`. View source, execute javascript, they're all done using `IBrowser` interface. Lots of common actions are actually extension methods see https://github.com/cefsharp/CefSharp/blob/cefsharp/96/CefSharp/WebBrowserExtensions.cs

> I am sorry that I don't know how to build the pre-version

If you looked a littler closer at https://github.com/cefsharp/CefSharp/issues/3902#issuecomment-978820002 you'd see the `-pre` release is available on `Nuget.org`


> like, if there's a private field of the ChromiumHostControl that refers to a chromium webBrowser or IWebBrowser

Doesn't really work that way.

In `96.0.142` I've made some additional improvements in commit https://github.com/cefsharp/CefSharp/commit/a152b2e6dba789707581169a592d30155e06b22c

`ChromiumHostControl` now shares a base interface with `ChromiumWebBrowser` called `IChromiumWebBrowserBase` which allows you to perform common operations. The `WinForms` example has also been updated so opening a popup in a tab will register basic events see https://github.com/cefsharp/CefSharp/blob/cefsharp/96/CefSharp.WinForms.Example/BrowserTabUserControl.cs#L35

**Comment by MeDeliog** [comment]
Hello, is there a way for the newly created ChromiumHostControl to have a corresponding JavascriptObjectRepository.ResolveObject ?
I'm using a new Windows Form with a ChromiumHostControl as a popup (not as a tab).

**Reply by amaitland** [reply]
> is there a way for the newly created ChromiumHostControl to have a corresponding JavascriptObjectRepository.ResolveObject ?

Not currently. Popups should still call JavascriptObjectRepository.ResolveObject for the parent ChromiumWebBrowser. You would need to use unique key for each object.
